build system: share rules to compile C files

shindere opened the pull request:
Several makefiles did contain rules to produce object files from
C files. This pull request gets rid of these essentially similar rules
and replaces all of them by one single rule, defined in `Makefile.common.in`.

Only one change can be observed when running `make world.opt`:
The `objinfo_helper` program is now compiled and linked in two distinct
steps rather than one before the PR.

Also, the changes in `otherlibs/systhreads/Makefile` and `stdlib/Makefile`
do not permit any sharing but make the compilation rules more uniform.

shindere commented:
This PR went through precheck on Inria's CI, see build #297.

gasche commented on otherlibs/systhreads/Makefile:
```diff
@@ -18,6 +18,10 @@ ROOTDIR=../..
 include $(ROOTDIR)/Makefile.config
 include $(ROOTDIR)/Makefile.common
 
+OC_CPPFLAGS += -I$(ROOTDIR)/runtime
```
You should also update the `.depend` target in this file, which can be simplified thanks to this definition.

gasche reviewed (approved):
I reviewed the change and believe it is correct, and I agree that it makes the makefiles a bit simpler: for files with non-standard flags, the new style lets us focus more on the non-standard flags.

shindere committed:
Build system: share rules to compile C files

This commit deduplicates some of the Makefile rules used to compile C files.

Rather than having one such rule per Makefile for each directory
containing C files, the relevant rules are moved to Makefile.common.in.

shindere commented:
Gabriel Scherer (2019/09/09 07:50 -0700):
> gasche approved this pull request.
>
> I reviewed the change and believe it is correct, and I agree that it
> makes the makefiles a bit simpler: for files with non-standard flags,
> the new style lets us focus more on the non-standard flags.

Thanks a lot for the prompt and positive feedback!

> > @@ -18,6 +18,10 @@ ROOTDIR=../..
>  include $(ROOTDIR)/Makefile.config
>  include $(ROOTDIR)/Makefile.common
>
> +OC_CPPFLAGS += -I$(ROOTDIR)/runtime
>
> You should also update the `.depend` target in this file, which can be
> simplified thanks to this definition.

Very well spotted, thanks. I did this change and went one step further
by introducing the `NATIVE_CPPFLAGS` variable.
